File: lib/python/openlinktoken/src/main/openlinktoken/metadata.py
# SPDX-License-Identifier: MIT

import hashlib
import sys
from typing import Any, Dict


class Metadata:
    """Handles metadata generation and management for Open Link Token."""

    # Metadata keys
    PLATFORM = "Platform"
    PYTHON_VERSION = "PythonVersion"
    VERSION = "Version"
    OUTPUT_FORMAT = "OutputFormat"
    BLANK_TOKENS_BY_RULE = "BlankTokensByRule"

    # Metadata values
    PLATFORM_PYTHON = "Python"
    METADATA_FILE_EXTENSION = ".metadata.json"
    SYSTEM_PYTHON_VERSION = f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"

    DEFAULT_VERSION = "2.2.0"

    # Output format values
    OUTPUT_FORMAT_JSON = "JSON"
    OUTPUT_FORMAT_CSV = "CSV"
    OUTPUT_FORMAT_PARQUET = "Parquet"

    def __init__(self):
        """Initialize the metadata object."""
        self.metadata_map: Dict[str, Any] = {}

    def initialize(self) -> Dict[str, Any]:
        """
        Initialize metadata with system information only.

        Returns:
            The initialized metadata map
        """
        self.metadata_map.clear()
        self.metadata_map[self.PYTHON_VERSION] = self.SYSTEM_PYTHON_VERSION
        self.metadata_map[self.PLATFORM] = self.PLATFORM_PYTHON
        self.metadata_map[self.VERSION] = self.DEFAULT_VERSION

        return self.metadata_map

    def add_hashed_secret(self, secret_key: str, secret_to_hash: str | bytes | None) -> Dict[str, Any]:
        """
        Set the secret and add its hash to the metadata.

        Args:
            secret_key: The key to use for the hashed secret in metadata
            secret_to_hash: The secret to hash

        Returns:
            The metadata map for method chaining
        """
        if isinstance(secret_to_hash, bytes):
            if secret_to_hash:
                self.metadata_map[secret_key] = self.calculate_secure_hash(secret_to_hash)
        elif secret_to_hash and secret_to_hash.strip():
            self.metadata_map[secret_key] = self.calculate_secure_hash(secret_to_hash)
        return self.metadata_map

    @staticmethod
    def calculate_secure_hash(input_str: str | bytes | None) -> str | None:
        """
        Calculate a secure SHA-256 hash of the given input.
        The hash is returned as a hexadecimal string, or ``None`` for empty input.

        Accepts both ``str`` (encoded as UTF-8) and raw ``bytes``.

        Args:
            input_str: The input string or raw bytes to hash.

        Returns:
            The SHA-256 hash as a hexadecimal string, or ``None`` when the input is empty.

        Raises:
            HashCalculationException: If there's an error calculating the hash.
        """
        if not input_str:
            return None

        try:
            hash_input = input_str if isinstance(input_str, bytes) else input_str.encode("utf-8")
            hash_object = hashlib.sha256(hash_input)
            hex_string = hash_object.hexdigest()
            return hex_string

        except Exception as e:
            raise HashCalculationException("Error calculating SHA-256 hash", e)


class HashCalculationException(Exception):
    """Custom exception for hash calculation errors."""

    def __init__(self, message: str, cause: Exception = None):
        super().__init__(message)
        self.cause = cause

File: lib/java/openlinktoken/src/main/java/org/openlinktoken/Metadata.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.LinkedHashMap;
import java.util.Map;

public class Metadata {

    // Metadata keys
    public static final String PLATFORM = "Platform";
    public static final String JAVA_VERSION = "JavaVersion";
    public static final String VERSION = "Version";
    public static final String OUTPUT_FORMAT = "OutputFormat";
    public static final String BLANK_TOKENS_BY_RULE = "BlankTokensByRule";

    // Metadata values
    public static final String PLATFORM_JAVA = "Java";
    public static final String METADATA_FILE_EXTENSION = ".metadata.json";
    public static final String SYSTEM_JAVA_VERSION = System.getProperty("java.version");

    public static final String DEFAULT_VERSION = "2.2.0";

    // Output format values
    public static final String OUTPUT_FORMAT_JSON = "JSON";
    public static final String OUTPUT_FORMAT_CSV = "CSV";
    public static final String OUTPUT_FORMAT_PARQUET = "Parquet";

    private Map<String, Object> metadataMap;

    public Metadata() {
        metadataMap = new LinkedHashMap<>();
    }

    /**
     * Initializes metadata with system information only.
     *
     * @return the initialized metadata map
     */
    public Map<String, Object> initialize() {
        metadataMap.clear();
        metadataMap.put(JAVA_VERSION, SYSTEM_JAVA_VERSION);
        metadataMap.put(PLATFORM, PLATFORM_JAVA);
        metadataMap.put(VERSION, DEFAULT_VERSION);

        return metadataMap;
    }

    /**
     * Sets the hashing secret and adds its hash to the metadata.
     *
     * @param secretToHash the secret to hash
     * @return the metadata map for method chaining
     */
    public Map<String, Object> addHashedSecret(String secretKey, String secretToHash) {
        if (secretToHash != null && !secretToHash.isEmpty()) {
            metadataMap.put(secretKey, calculateSecureHash(secretToHash));
        }
        return metadataMap;
    }

    /**
     * Sets a raw-byte secret and adds its hash to the metadata.
     *
     * @param secretToHash the raw secret bytes to hash
     * @return the metadata map for method chaining
     */
    public Map<String, Object> addHashedSecret(String secretKey, byte[] secretToHash) {
        if (secretToHash != null && secretToHash.length > 0) {
            metadataMap.put(secretKey, calculateSecureHash(secretToHash));
        }
        return metadataMap;
    }

    /**
     * Calculates a secure SHA-256 hash of the given input.
     * The hash is returned as a hexadecimal string.
     *
     * @param input the input string to hash
     * @return the SHA-256 hash as a hexadecimal string
     * @throws HashCalculationException if SHA-256 algorithm is not available
     */
    public static String calculateSecureHash(String input) {
        if (input == null || input.isEmpty()) {
            return null;
        }

        return calculateSecureHash(input.getBytes(StandardCharsets.UTF_8));
    }

    /**
     * Calculates a secure SHA-256 hash of the given raw bytes.
     * The hash is returned as a hexadecimal string.
     *
     * @param input the input bytes to hash
     * @return the SHA-256 hash as a hexadecimal string
     * @throws HashCalculationException if SHA-256 algorithm is not available
     */
    public static String calculateSecureHash(byte[] input) {
        if (input == null || input.length == 0) {
            return null;
        }

        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hashBytes = digest.digest(input);

            // Convert bytes to hexadecimal string
            StringBuilder hexString = new StringBuilder();
            for (byte b : hashBytes) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) {
                    hexString.append('0');
                }
                hexString.append(hex);
            }
            return hexString.toString();

        } catch (NoSuchAlgorithmException e) {
            throw new HashCalculationException("SHA-256 algorithm not available", e);
        }
    }

    /**
     * Custom exception for hash calculation errors.
     */
    public static class HashCalculationException extends RuntimeException {
        public HashCalculationException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}
